rpm/0070-sailfishos-webrtc-Implement-video-capture-module.-JB.patch

```diff
@@ -0,0 +1,2255 @@
++    if (!_mainLoop) {
++        gst_bus_remove_signal_watch (_bus);
++        gst_object_unref (_bus);
++        gst_object_unref (_capsFilterDimensions);
++        gst_object_unref (_capsFilterRate);
++        gst_object_unref (_pipeline);
++        goto init_error;
++    }
++
++    ret = 0;
++init_error:
++    g_main_context_pop_thread_default (_captureContext);
++  
++    return ret;
++}
++
++GstDroidCamera::~GstDroidCamera()
++{
++    StopCapture();
++
++    gst_element_set_state (_pipeline, GST_STATE_NULL);
++
++    gst_bus_remove_signal_watch(_bus);
++    gst_object_unref (_bus);
++    gst_object_unref (_capsFilterDimensions);
++    gst_object_unref (_capsFilterRate);
++    gst_object_unref (_pipeline);
++    g_main_loop_unref(_mainLoop);
++    g_main_context_unref(_captureContext);
++
++    LOG_F(LS_VERBOSE);
++}
++
++int32_t GstDroidCamera::StartCapture(
++    const VideoCaptureCapability& capability)
++{
++    GstCaps *caps;
++    gchar caps_string[128];
++
++    LOG_F(LS_VERBOSE);
++
++    if (_captureStarted) {
++        if (capability.width == _currentCaptureSettings.width &&
++        capability.height == _currentCaptureSettings.height &&
++        capability.maxFPS == _currentCaptureSettings.maxFPS &&
++        capability.rawType == _currentCaptureSettings.rawType) {
++            return 0;
++        } else {
++            StopCapture();
++        }
++    }
++
++    /* Configure frame size */
++    g_snprintf (caps_string, sizeof(caps_string),
++            "video/x-raw(memory:DroidMediaQueueBuffer), width=%u, height=%u",
++            capability.width, capability.height);
++
++    caps = gst_caps_from_string (caps_string);
++    if (caps) {
++        LOG_F(LS_VERBOSE) << "Applying caps: " << caps_string;
++        g_object_set (G_OBJECT (_capsFilterDimensions), "caps", caps, NULL);
++    } else {
++        LOG_F(LS_ERROR) << "Cannot set image resolution";
++    }
++    gst_caps_unref (caps);
++
++    /* Configure frame rate */
++    g_snprintf (caps_string, sizeof(caps_string),
++            "video/x-raw(memory:DroidMediaQueueBuffer), framerate=%u/1",
++            capability.maxFPS);
++
++    caps = gst_caps_from_string (caps_string);
++    if (caps) {
++        LOG_F(LS_VERBOSE) << "Applying caps: " << caps_string;
++        g_object_set (G_OBJECT (_capsFilterRate), "caps", caps, NULL);
++    } else {
++        LOG_F(LS_ERROR) << "Cannot set frame rate";
++    }
++    gst_caps_unref (caps);
++
++    _currentCaptureSettings.rawType = kVideoUnknown;
++    _currentCaptureSettings.width = capability.width;
++    _currentCaptureSettings.height = capability.height;
++    _currentCaptureSettings.maxFPS = capability.maxFPS;
++    /* FIXME: Is it possible to read capture delay from the HW? */
++    _currentCaptureSettings.expectedCaptureDelay =
++            mozilla::Preferences::GetUint(EXPECTED_CAPTURE_DELAY_PREF, 500);
++
++    /* Pre-calculate frame size for YV12 */
++    _frameSize = capability.width * capability.height * 3 / 2;
```

Comment:
This size may be too small if the camera source adds some padding to the row width for memory alignment reasons.

gst-droid calculates its buffer sizes here https://github.com/sailfishos/gst-droid/blob/master/gst-libs/gst/droid/gstdroidmediabuffer.c#L260. I'm not sure how much of the info can be extracted from gstreamer ahead of time, but it should become available when you map the gstreamer buffer.